xbmc/cores/VideoPlayer/MediaPipelineWebOS.h

```diff
@@ -0,0 +1,440 @@
+  std::unique_ptr<CAEEncoderFFmpeg> m_audioEncoder{nullptr};
+  std::atomic<int> m_bufferLevel{0};
+  std::atomic<bool> m_audioFull{false};
+  std::atomic<bool> m_videoFull{false};
+  std::atomic<unsigned long> m_droppedFrames{0};
+
+  std::mutex m_audioCriticalSection;
+
+  CDVDMessageQueue m_messageQueueAudio;
+  CDVDMessageQueue m_messageQueueVideo;
+  CDVDMessageQueue& m_messageQueueParent;
+  CProcessInfo& m_processInfo;
+  CRenderManager& m_renderManager;
+  CDVDClock& m_clock;
+  CDVDOverlayContainer& m_overlayContainer;
+  bool m_hasAudio{true};
+
+  std::string m_audioInfo;
+  std::string m_videoInfo;
+  BitstreamStats m_audioStats{};
+  BitstreamStats m_videoStats{};
+
+  std::thread m_audioThread;
+};
+
+/**
+ * @class CVideoPlayerVideoWebOS
+ * @brief Video stream player adapter forwarding to CMediaPipelineWebOS.
+ */
+class CVideoPlayerVideoWebOS final : public IDVDStreamPlayerVideo
+{
+public:
+  CVideoPlayerVideoWebOS(CMediaPipelineWebOS& mediaPipeline, CProcessInfo& processInfo)
+    : IDVDStreamPlayerVideo(processInfo), m_mediaPipeline(mediaPipeline)
+  {
+  }
+  void FlushMessages() override { m_mediaPipeline.FlushVideoMessages(); }
+  bool OpenStream(const CDVDStreamInfo hint) override
+  {
+    return m_mediaPipeline.OpenVideoStream(hint);
+  }
+  void CloseStream(const bool waitForBuffers) override
+  {
+    m_mediaPipeline.CloseVideoStream(waitForBuffers);
+  }
+  void Flush(const bool sync) override { m_mediaPipeline.Flush(sync); }
+  [[nodiscard]] bool AcceptsData() const override { return m_mediaPipeline.AcceptsVideoData(); }
+  [[nodiscard]] bool HasData() const override { return m_mediaPipeline.HasVideoData(); }
+  [[nodiscard]] bool IsInited() const override { return m_mediaPipeline.IsVideoInited(); }
+  void SendMessage(const std::shared_ptr<CDVDMsg> msg, const int priority) override
+  {
+    m_mediaPipeline.SendVideoMessage(msg, priority);
+  }
+  void EnableSubtitle(const bool enable) override { m_mediaPipeline.EnableSubtitle(enable); }
+  bool IsSubtitleEnabled() override { return m_mediaPipeline.IsSubtitleEnabled(); }
+  double GetSubtitleDelay() override { return m_mediaPipeline.GetSubtitleDelay(); }
+  void SetSubtitleDelay(const double delay) override { m_mediaPipeline.SetSubtitleDelay(delay); }
+  [[nodiscard]] bool IsStalled() const override { return m_mediaPipeline.IsStalled(); }
+  double GetCurrentPts() override { return m_mediaPipeline.GetCurrentPts(); }
+  double GetOutputDelay() override { return 0.0; }
+  std::string GetPlayerInfo() override { return m_mediaPipeline.GetVideoInfo(); }
+  int GetVideoBitrate() override { return m_mediaPipeline.GetVideoBitrate(); }
+  void SetSpeed(const int speed) override { m_mediaPipeline.SetSpeed(speed); }
+
+private:
+  CMediaPipelineWebOS& m_mediaPipeline;
+};
+
+/**
+ * @class CVideoPlayerAudioWebOS
+ * @brief Audio stream player adapter forwarding to CMediaPipelineWebOS.
+ */
+class CVideoPlayerAudioWebOS final : public IDVDStreamPlayerAudio
```

Comment:
Can separate classes be split into separate files? I got a little scared when I saw a 440 line header, not realizing it was multiple classes.

Reply:
Sure but those two classes mostly just forward into the pipeline class.

Reply:
OK.

Reply:
They live in their own header files now. Do you think that's alright or should I also split the definitions into their own cpp files?

Reply:
I'd split. Files are cheap, and long headers should be avoided. IMO it also helps to see the architecture at a FS level before diving into the code.

Reply:
Done, CLions split function definition is great